ambry-router/src/main/java/com.github.ambry.router/AdaptiveOperationTracker.java

```diff
@@ -77,15 +77,17 @@
   }
 
   @Override
-  public void onResponse(ReplicaId replicaId, boolean isSuccessFul) {
-    super.onResponse(replicaId, isSuccessFul);
+  public void onResponse(ReplicaId replicaId, boolean isSuccessFul, RouterErrorCode routerErrorCode) {
```

Comment:
How about getting rid of the `boolean` in that case and determine success or failure purely based on the error code?

Reply:
It is viable but a little aggressive in this PR.  I see your points here, instead of relying on the caller, the operation tracker should resolve the `isSuccessful` status by itself. Personally, I am a little conservative on this, maybe we can remove that in the future PR. What do you think?

Reply:
On second thought, you may not be able to determine success or failure based on the error code alone, as the interpretation might be different for different operation types (for example, `Blob_Deleted` is an error for a Get, but perhaps not for a Delete operation). It would still be nice if the signature can be made more elegant, but I dont' have a good suggestion at this point.

Reply:
Perhaps we could create another enum to represent how the operation tracker should handle the response, but abstracts out the actual error code.

e.g.: `enum OperationTracker.Status {SUCCESS, FAILURE, TIMED_OUT }`

Reply:
sounds like a good idea. Let me make the changes as suggested.

Reply:
Instead using of `OperationTracker.Status`, I decide to use the  `RequestResult` here. It is more reasonable in context of `onResponse` method. Also `RequestResult` is more general, it includes the case where request is not successfully sent out due to connection checkout timeout etc.

Reply:
`RequestResult ` is a very common name. A more descriptive name would be good. 

Also it would be great if add some comments about the case @pnarayanan mentioned, to remind us in the future.

Reply:
changed the enum name to `RouterRequestFinalState`